certbot/certbot/_internal/client.py

```diff
@@ -528,6 +534,22 @@ def obtain_and_enroll_certificate(self, domains: List[str], certname: Optional[s
             key.pem, chain,
             self.config)
 
+    def _retry_obtain_certificate(self, error: messages.Error, key: util.Key,
+                                csr: util.CSR, domains: List[str],
+                                ) -> Tuple[bytes, bytes, util.Key, util.CSR]:
+        if self.config.allow_subset_of_names and error.subproblems is not None:
+            failed_domains = [problem.identifier.value for problem in error.subproblems
+                                if problem.identifier is not None]
+            successful_domains = [x for x in domains if x not in failed_domains]
+        if successful_domains != domains and len(successful_domains) != 0:
+            display_util.notify("Unable to obtain a certificate with every requested "
+                "domain. Retrying without: {0}".format(failed_domains))
+            if not self.config.dry_run:
+                os.remove(key.file)
+                os.remove(csr.file)
+            return self.obtain_certificate(successful_domains)
```

Comment:
ISTM that this second `if` branch should be nested within the first one.

As written: if the ACME server rejects an identifier but does not include any `subproblems`, the function will crash.

It would be good to include a test case for this kind of server response as well.

Reply:
Thanks - I hadn't spotted that case. I've moved the second `if` statement, and added tests for an acme error during `_get_order_and_authorizations` and `obtain_certificate_from_csr` with no subproblems.

Before applying the changes to `client.py`, the new tests fail due to `successful_domains` being accessed before it is assigned, because the first `if` statement evaluates to false and the second evaluates to true.

After applying the changes to `client.py`, the new tests pass, and the returned acme error is raised without retrying.